HDDS-12553. ozone admin container list should output real JSON array

kerneltime opened the pull request:
## What changes were proposed in this pull request?

Support listing containers in json array format. 

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-12553
## How was this patch tested?

Additional robot tests.

errose28 commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java:
```diff
@@ -90,11 +95,35 @@ public class ListSubcommand extends ScmSubcommand {
+    // JSON array format with proper brackets and commas
+    if (isFirst) {
+      // Start of array
+      System.out.print("[");
```
Let's use Jackson's `SequenceWriter` for this similar to what is done in #7944. We may need to duplicate some minor changes to `JsonUtils` here but that should be ok and we can resolve it on the next reverse merge into reconciler branch.

kerneltime replied:
Will look into SequenceWriter

errose28 commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java:
```diff
@@ -76,6 +76,11 @@ public class ListSubcommand extends ScmSubcommand {
   private String replication;
+  
+  @Option(names = {"--json"},
+      description = "Output the entire list in JSON array format",
```
The original is still json (and can still be streamed into jq), it just doesn't represent a valid json file if it is saved that way. In that sense I think the flag here is confusing. I would just make json list the default and only behavior. We did something similar in HDDS-5824 and did not consider it a breaking change, since the original was a "bug". With 2.0 coming up such a change makes sense as well.

errose28 replied:
Existing acceptance tests would need to be updated to account for this.

kerneltime replied:
Yeah, I was trying to avoid changing the existing behavior of the CLI, as it would be good to tread CLI output as a contract. I ran into JSON parsing errors when analyzing a bunch of returned containers. `jq` works fine with streaming in the container JSONs, but it would be good to have JSON array as per speck as well.

errose28 commented on hadoop-ozone/dist/src/main/smoketest/admincli/container.robot:
```diff
@@ -96,6 +96,33 @@ List all containers from a particular container ID
 
+List containers in JSON array format
+    ${output} =         Execute          ozone admin container list --json | jq -r '.'
+                        Should Start With   ${output}   [
```
This way of testing works, but a jq query to do the same test might be more robust (although failure messages may be more confusing):
`ozone admin container list --json | jq '.[] | .containerID'` then ensure that we actually get content in the output.

kerneltime replied:
The goal is to see if the parsing matches, other tests delve deeper into the returned payload. I can add some additional checks here as well.

adoroszlai replied:
> `ozone admin container list --json | jq '.[] | .containerID'`

I suggest splitting it into two commands:

```
${output} =     Execute    <command returns JSON>
${excerpt} =    Execute    echo '${output}' | jq '....'
```

This has the benefit of showing raw output, which helps in case `jq` fails to find expected structure.  Plus you can reuse `output` for multiple extractions.

errose28 reviewed (commented):
Thanks for fixing this @kerneltime.

kerneltime commented:
@errose28 @adoroszlai thank you for the review, I have addressed the review comments.

kerneltime commented:
@errose28 for now I am not breaking the old CLI output, if developers feel strongly we can do it as a separate PR with updates to existing tests that fail.

adoroszlai commented:
I know duplication is temporary, but it still bothers me.

```diff
diff --git hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java
index 888a01f211..c651750ce3 100644
--- hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java
+++ hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java
@@ -26,6 +26,7 @@
 import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
 import com.google.common.base.Strings;
 import java.io.IOException;
+import java.util.List;
 import org.apache.hadoop.hdds.cli.HddsVersionProvider;
 import org.apache.hadoop.hdds.client.ReplicationConfig;
 import org.apache.hadoop.hdds.client.ReplicationType;
@@ -36,6 +37,8 @@
 import org.apache.hadoop.hdds.scm.client.ScmClient;
 import org.apache.hadoop.hdds.scm.container.ContainerInfo;
 import org.apache.hadoop.hdds.scm.container.ContainerListResult;
+import org.apache.hadoop.hdds.utils.IOUtils;
+import org.apache.ratis.util.function.CheckedConsumer;
 import picocli.CommandLine.Command;
 import picocli.CommandLine.Help.Visibility;
 import picocli.CommandLine.Option;
@@ -96,7 +99,7 @@ public class ListSubcommand extends ScmSubcommand {
   }
 
 
-  private void outputContainerInfo(ContainerInfo containerInfo) throws IOException {
+  private static void outputContainerInfo(ContainerInfo containerInfo) throws IOException {
     // Original behavior - just print the container JSON
     System.out.println(WRITER.writeValueAsString(containerInfo));
   }
@@ -120,6 +123,7 @@ public void execute(ScmClient scmClient) throws IOException {
             ScmConfigKeys.OZONE_SCM_CONTAINER_LIST_MAX_COUNT_DEFAULT);
 
     ContainerListResult containerListAndTotalCount;
+    ContainerListProcesor processor = jsonFormat ? new JsonListWriter() : new JsonItemsWriter();
 
     if (!all) {
       if (count > maxCountAllowed) {
@@ -129,19 +133,11 @@ public void execute(ScmClient scmClient) throws IOException {
         count = maxCountAllowed;
       }
       containerListAndTotalCount = scmClient.listContainer(startId, count, state, type, repConfig);
-      
-      if (jsonFormat) {
-        // JSON array format using SequenceWriter directly
-        try (SequenceWriter sequenceWriter = WRITER.writeValues(System.out)) {
-          sequenceWriter.init(true); // Initialize as a JSON array
-          sequenceWriter.writeAll(containerListAndTotalCount.getContainerInfoList());
-        }
-        System.out.println(); // Add final newline
-      } else {
-        // Original format - one JSON object per line
-        for (ContainerInfo container : containerListAndTotalCount.getContainerInfoList()) {
-          outputContainerInfo(container);
-        }
+
+      try {
+        processor.accept(containerListAndTotalCount.getContainerInfoList());
+      } finally {
+        IOUtils.closeQuietly(processor);
       }
 
       if (containerListAndTotalCount.getTotalCount() > count) {
@@ -154,46 +150,61 @@ public void execute(ScmClient scmClient) throws IOException {
       int batchSize = (count > 0) ? count : maxCountAllowed;
       long currentStartId = startId;
       int fetchedCount;
-      
-      if (jsonFormat) {
-        // For JSON array format, use SequenceWriter to stream containers as we fetch them
-        try (SequenceWriter sequenceWriter = WRITER.writeValues(System.out)) {
-          sequenceWriter.init(true); // Initialize as a JSON array
-          
-          do {
-            // Fetch containers in batches of 'batchSize'
-            containerListAndTotalCount = scmClient.listContainer(currentStartId, batchSize, state, type, repConfig);
-            fetchedCount = containerListAndTotalCount.getContainerInfoList().size();
-            
-            // Write containers directly to the SequenceWriter
-            for (ContainerInfo container : containerListAndTotalCount.getContainerInfoList()) {
-              sequenceWriter.write(container);
-            }
-            
-            if (fetchedCount > 0) {
-              currentStartId =
-                  containerListAndTotalCount.getContainerInfoList().get(fetchedCount - 1).getContainerID() + 1;
-            }
-          } while (fetchedCount > 0);
-        }
-        System.out.println(); // Add final newline
-      } else {
-        // Original format - one JSON object per line
+
+      try {
         do {
           // Fetch containers in batches of 'batchSize'
           containerListAndTotalCount = scmClient.listContainer(currentStartId, batchSize, state, type, repConfig);
           fetchedCount = containerListAndTotalCount.getContainerInfoList().size();
 
-          for (ContainerInfo container : containerListAndTotalCount.getContainerInfoList()) {
-            outputContainerInfo(container);
-          }
+          // Write containers directly to the SequenceWriter
+          processor.accept(containerListAndTotalCount.getContainerInfoList());
 
           if (fetchedCount > 0) {
             currentStartId =
                 containerListAndTotalCount.getContainerInfoList().get(fetchedCount - 1).getContainerID() + 1;
           }
         } while (fetchedCount > 0);
+      } finally {
+        IOUtils.closeQuietly(processor);
       }
     }
   }
+
+  private interface ContainerListProcesor extends CheckedConsumer<List<ContainerInfo>, IOException>, AutoCloseable {
+    //
+  }
+
+  private static final class JsonListWriter implements ContainerListProcesor {
+    private final SequenceWriter sequenceWriter;
+
+    private JsonListWriter() throws IOException {
+      // For JSON array format, use SequenceWriter to stream containers as we fetch them
+      sequenceWriter = WRITER.writeValues(System.out).init(true);
+    }
+
+    @Override
+    public void accept(List<ContainerInfo> list) throws IOException {
+      sequenceWriter.writeAll(list);
+    }
+
+    @Override
+    public void close() throws Exception {
+      sequenceWriter.close();
+    }
+  }
+
+  private static final class JsonItemsWriter implements ContainerListProcesor {
+    @Override
+    public void accept(List<ContainerInfo> list) throws IOException {
+      for (ContainerInfo container : list) {
+        outputContainerInfo(container);
+      }
+    }
+
+    @Override
+    public void close() throws Exception {
+      System.out.println(); // Add final newline
+    }
+  }
 }
```

kerneltime commented:
Hmm let me just get rid of the backward compatibility, as it was a bug to begin with.

kerneltime commented:
@adoroszlai why prefer `IOUtils.closeQuietly` ?

adoroszlai commented:
> why prefer `IOUtils.closeQuietly` ?

Convenience.  I don't expect `SequenceWriter.close` to have any problem, but implementing `AutoCloseable` requires declaring `throws Exception`.

kerneltime commented:
> > why prefer `IOUtils.closeQuietly` ?
> 
> Convenience. I don't expect `SequenceWriter.close` to have any problem, but implementing `AutoCloseable` requires declaring `throws Exception`.
> 
> Using `try-with` is also fine, probably could also rewrite my suggestion that way.

Yeah.. I think I will

kerneltime commented:
I am facing an issue with the integration test; for some reason, `out` does not reflect up the output of the second command in the integration tests, and this happens after I switch to sequenceWriter... still debugging.

kerneltime commented:
Found the issue the absence of new line was throwing off the integration tests. Fixed that and also added additional tests to check that a new line should be printed post the json array.

adoroszlai commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/scm/cli/container/ListSubcommand.java:
```diff
@@ -114,44 +116,104 @@ public void execute(ScmClient scmClient) throws IOException {
+  }
+
+  private void listAllContainers(ScmClient scmClient, SequenceWriter writer,
+                                 int batchSize, ReplicationConfig repConfig)
```
nit: Please do not format method signature like this. Whenever visibility / return type / method name / other modifiers are changed, we would have to reindent parameters.

kerneltime replied:
I can change this in a followup.

adoroszlai reviewed (commented):
Thanks @kerneltime for updating the patch.  I'm fine with merging as is, and handling some minor improvements in follow-up.

kerneltime commented on hadoop-ozone/dist/src/main/smoketest/admincli/container.robot:
```diff
@@ -87,17 +97,54 @@ Report containers as JSON
+                        Should Not Contain   ${states}   CLOSED
+
+Check count limit with JSON array format
+    ${output} =         Execute          ozone admin container create
+                        Should contain   ${output}   is created
+    ${output} =         Execute          ozone admin container create
+                        Should contain   ${output}   is created
+    ${output} =         Execute          ozone admin container create
+                        Should contain   ${output}   is created
+    ${output} =         Execute          ozone admin container create
+                        Should contain   ${output}   is created
+    ${output} =         Execute          ozone admin container create
+                        Should contain   ${output}   is created
```
Yes the robot tests need an over haul

adoroszlai commented:
Thanks @kerneltime for the patch, @errose28 for the review.

Please make sure to file follow-up tasks.

kerneltime commented:
> Thanks @kerneltime for the patch, @errose28 for the review.
> 
> Please make sure to file follow-up tasks.

Done https://issues.apache.org/jira/browse/HDDS-12743
